### ISSUE TITLE
pachd opens way too many gRPC connections

### ISSUE THREAD

**Issue by derekchiang** [pachyderm/pachyderm#588]
This happens when you write a lot of files.  The number of ports used can shoot up to 20k+.  It also causes a significant amount of memory usage.  For instance, over 5GB of memory is allocated by gRPC over the course of writing 10000+ files.

```
      flat  flat%   sum%        cum   cum%
191164050.63kB 95.74% 95.74% 191164050.63kB 95.74%  github.com/pachyderm/pachyderm/src/server/vendor/bazil.org/fuse.allocMessage
5145702.55kB  2.58% 98.32% 5179682.23kB  2.59%  github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc/transport.newFramer
```

**Comment by derekchiang** [comment]
An update on this.  gRPC opens a new TCP connection whenever a RPC is invoked.  This in itself is not problematic, since the port will be reclaimed after the RPC is finished.  However, the way TCP works, closed connections can remain in the TIME-WAIT state for a few tens of seconds, waiting for out-of-order packets.  Since we are making RPCs very rapidly, we are observing a large amount of TCP connections in the TIME-WAIT state.

![](https://d1g3mdmxf8zbo9.cloudfront.net/images/tcp/tcp-state-diagram.png)

However, there is currently no evidence that this is a problem.  The aforementioned high memory usage is mostly caused by GC, as described [here](https://github.com/pachyderm/pachyderm/issues/581).

A related issue: https://github.com/grpc/grpc/issues/5582

**Comment by tv42** [comment]
Not sure why you closed this. gRPC is supposed to send multiple RPCs over one connection, even concurrently.

A quick `tcpdump` says that a `TestWriteManyFiles` causes 40040 connections to be opened. That's clearly way, way, too many.

**Comment by jdoliner** [comment]
I'm kind of confused how we're opening so many connections. It looks like we pass a single connection into `NewMounter` (wrapped in a `PfsAPIClient`). Is it possible that streaming rpcs result in a new connection being opened?

**Comment by tv42** [comment]
A lot of code seems to look like

```go
		clientConn, err := a.getClientConnForFile(request.File, a.version)
		if err != nil {
			return err
		}
		defer clientConn.Close()
```

as clarified on slack, apparently grpc connection reuse was explicitly taken out (for now) because grpc had trouble recovering from a hung TCP connection. Well, too many connections is the other end of the spectrum :(

Expect random test & benchmark failures, expect weird drops in sustained throughput.

**Comment by derekchiang** [comment]
Reopening and renaming this issue to reflect the new finding that it's the pachd servers that are opening too many connections, as @tv42 pointed out.  We haven't run into a noticeable problem with that yet, but it's definitely very much less than ideal.

**Comment by jdoliner** [comment]
As mentioned offline, we actually don't need to solve this issue directly. We're planning a refactor of pfs in #411 which will make it so that pachd nodes no longer need to connect to each other which sidesteps this problem.

It would still be really nice if we had a reliable way for long running processes to keep grpc connections to one another... but alas I don't see how to do it right now.

Moving this issue to 1.2 which is where we're planning to do the refactor.

**Comment by tv42** [comment]
It seems https://godoc.org/google.golang.org/grpc#WithDialer with a dialer that sets TCP keepalive with https://golang.org/pkg/net/#TCPConn.SetKeepAlive would help the problem of hung gRPC connections.

**Comment by derekchiang** [comment]
@tv42 yup, we have since realized that gRPC does try to reconnect by default.  Just merged #591 which allows us to reuse connections.
